[iOS] Full keyboard access scrolling

LongCatIsLooong opened the pull request:
This PR adds basic FKA scrolling support: when the iOS focus (the focus state is maintained separately from the framework focus, see the previous PR) switches to an item in a scrollable container that is too close to the edge of the viewport, the container will scroll to make sure the next item is visible. 

Previous PR for context: https://github.com/flutter/engine/pull/55964

https://github.com/user-attachments/assets/84ae5153-f955-4d23-9901-ce942c0e98ac


### Why the UIScrollView subclass in the focus hierarchy

The iOS focus system does not provide an API that allows apps to notify it of focus highlight changes. So if we were to keep using the transforms sent by the framework as-is and not introducing any UIViews in the focus hierarchy, the focus highlight will be positioned at the wrong location after scrolling (via FKA or via framework). That does not seem to be part of the public API and the focus system seems to only know how to properly highlight focusable UIViews.

### Things that currently may not work

1. Nested scroll views (have not tried to verify) 

The `UIScrollView`s are always subviews of the `FlutterView`. If there are nested scrollables the focus system may not be able to properly determine the focus hierarchy (in theory the iOS focus system should never depend on `UIView.parentView` but I haven't tried to verify that).

2. If the next item is too far below the bottom of the screen and there is a tab bar with focusable items, the focus will be transferred to tab bar instead of the next item in the list

Video demo (as you can see the scrolling is really finicky):

https://github.com/user-attachments/assets/51c2bfe4-d7b3-4614-aa49-4256214f8978

I've tried doing the same thing using a `UITableView` with similar configurations but it seems to have the same problem. I'll try to dig a bit deeper into this and see if there's a workaround.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

LongCatIsLooong commented on shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:
```diff
@@ -174,7 +176,9 @@ - (void)accessibilityBridgeDidFinishUpdate {
   self.scrollView.frame = self.accessibilityFrame;
   self.scrollView.contentSize = [self contentSizeInternal];
-  [self.scrollView setContentOffset:[self contentOffsetInternal] animated:NO];
+  if (!self.scrollView.isDoingSystemScrolling) {
```
This means if the scroll view is in a FKA scrolling animation, all framework updates will be disregarded. 

This is not optimal, if the user wants to interrupt the scrolling animation via swipe gestures, the `UIScrollView` will be temporarily out of sync with the framework scroll offset, but once the animation is done, this `UIScrollView` will send one last offset update to the framework, so it's guaranteed that the framework semantics tree will need another update and the new tree will also update the accessibility bridge and the offsets maintained by the scroll view and the framework will become consistent again.

LongCatIsLooong replied:
I have tried removing this flag, and the offsets echoed back by the framework seem always to be one step behind and I am not sure exactly why:

It seems `Shell::OnPlatformViewDispatchSemanticsAction` and `Shell::OnEngineUpdateSemantics` both run tasks using `RunNowOrPostTask`. Now that the UI thread is the same as the platform thread, there should be no delay, I was expecting:
```
<<<FKA setContentOffset: {0, 13.333333333333334}
<<<FKA setContentOffset: {0, 13.333333333333334}
>>>framework setContentOffset: {0, 13.333333333333343}
<<<FKA setContentOffset: {0, 15}
<<<FKA setContentOffset: {0, 15}
>>>framework setContentOffset: {0, 15}
```

but instead it's always delayed:
```
<<<FKA setContentOffset: {0, 13.333333333333334}
<<<FKA setContentOffset: {0, 13.333333333333334}
... 
<<<FKA setContentOffset: {0, 15}
<<<FKA setContentOffset: {0, 15}
>>>framework setContentOffset: {0, 15}
```

chunhtai replied:
yes, i think this makes sense since FKA set content will change the scrollcontroller, but it will have to wait for one frame to get the layout and semantics update back

cbracken replied:
Please add a comment covering this here.

LongCatIsLooong replied:
So it looks like I was looking at an older checkout, and it was changed [here](https://github.com/flutter/engine/pull/56514). The event ordering becomes the way I expected it to be if we run `DispatchSemanticsAction` synchronously like before. Jonah said `RunNowOrPostTask` semantics action dispatch is fine we'll just need to post an empty task afterwards to flush the dart futures. I think if we do that instead we could eliminate this flag.

That seems to be the better option because we don't need the flag and the scroll view will be responsive to user gestures during the FKA scroll animation, but I think I'll make the attempt a separate PR.

LongCatIsLooong replied:
> I am curious to see if it will work, but I think even if the action is send to framework synchronously, it may still take one frame for flutter to rebuild layout and flush semantics back

That's fine I think. The problem with `DispatchSemanticsAction` being async is that we are essentially maintaining two sources of truth so it becomes a distributed system where the communication channels only guarantee FIFO ordering. If the UIScrollView here and the framework semantics tree having different opinion about what the content offset should be, the update takes time to travel and the inconsistency can be observed. 

If the communication is synchronous then the inconsistent state won't be observable as the updates would arrive instantaneously.

reidbaker reviewed (commented):
android code LGTM

chunhtai commented on shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm:
```diff
@@ -48,6 +55,9 @@ - (void)didUpdateFocusInContext:(UIFocusUpdateContext*)context
 }
 
 - (id<UIFocusEnvironment>)parentFocusEnvironment {
+  if ([self.parent isKindOfClass:[FlutterScrollableSemanticsObject class]]) {
```
doesn't this need to be recursively up?

chunhtai replied:
since we recursively looks up in frame calculation

LongCatIsLooong replied:
What do you mean by recursively up? This method is supposed to return the parent object in the focus hierarchy not the root (this method is part of the `UIFocusEnvironment` protocol and the iOS focus engine will use this method to traverse the focus hierarchy).

chunhtai commented on shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm:
```diff
@@ -71,8 +81,56 @@ - (BOOL)canBecomeFocused {
+// The frame is described in the `coordinateSpace` of the
+// `parentFocusEnvironment` (all `parentFocusEnvironment`s are `UIFocusItem`s).
+//
+// Seealso the `coordinateSpace` implementation.
```
```suggestion
// See also the `coordinateSpace` implementation.
```

LongCatIsLooong replied:
Ha I thought seealso was a word.

chunhtai commented on shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm:
```diff
@@ -87,16 +145,99 @@ - (CGRect)frame {
   //
   // This method is only supposed to return items within the given
```
just a note that I have run into weird bug before because i returned child something outside of the rect of accessibilityContainer. I am not sure if it applies here.

LongCatIsLooong replied:
Yeah I can try adding culling here but I'll have to test it separately (and it seems to be working fine w/o the culling).

chunhtai commented on shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm:
```diff
@@ -87,16 +145,99 @@ - (CGRect)frame {
+  // to a Float64List in dart.
+  // The size of a CGFloat is architecture-dependent and it's typically the word
+  // size. The last iOS with 32-bit support was iOS 10.
+  static_assert(sizeof(CGPoint) == sizeof(CGFloat) * 2);
```
should we make it a soft-crash or short circuit if this condition does not hold

LongCatIsLooong replied:
This is a compile-time check. We typically deliver the engine as a prebuilt binary so it's not going to a concern for most users / devs. This assert is mostly for code readers to see the assumptions we made about CGFloat here.

cbracken replied:
This feels pretty naughty. Even just for the sake of readability, I'd be tempted to create a `double[]`, copy the values in, then encode and ship that. The extra CPU work is negligible and the one line of code will save you five lines of comment.

cbracken commented on shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm:
```diff
@@ -3,7 +3,10 @@
 // found in the LICENSE file.
```
Despite how it looks, yep. This is the correct name format for categories on existing classes.

chunhtai replied:
thanks, TIL.

LongCatIsLooong commented on shell/platform/darwin/ios/framework/Source/FlutterSemanticsScrollView.h:
```diff
@@ -18,10 +18,30 @@ NS_ASSUME_NONNULL_BEGIN
+/// This flag is set by the `FlutterSemanticsScrollView` itself, typically in
+/// one of the `UIScrollViewDelegate` methods.
+///
+/// When this flag is true, the Flutter framework should typically cease
```
The flag is only read by the engine code and the framework side does not know about this. Yeah the wording is confusing I'll change it to "When this flag is true, the SemanticsObject ignores all incoming framework updates"

LongCatIsLooong replied:
(updated the comment)

cbracken commented on shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm:
```diff
@@ -71,8 +81,56 @@ - (BOOL)canBecomeFocused {
+  }
+
+  SkRect rect;
+  rect.setBounds(quad, 4);
```
I'm presuming the reason for using Skia types in the first half of the methods is that `self.node.transform` was already a SkM44? That seems reasonable given the existing code but would you mind adding a TODO to migrate this code to CoreGraphics types? The less Skia-dependent code we can add in the embedders themselves, the better.

LongCatIsLooong replied:
I basically copied this from https://github.com/flutter/engine/blob/764b91d7d75af08a334a1509cd24cdefcc16f2e5/shell/platform/darwin/ios/framework/Source/SemanticsObject.mm#L66-L95. 

But here `setBounds` is probably just taking the `MIN` and `MAX` of the coordinates. I'll do that by hand.

cbracken commented on shell/platform/darwin/ios/framework/Source/SemanticsObjectTest.mm:
```diff
@@ -685,6 +691,10 @@ - (void)testFlutterScrollableSemanticsObjectHidesScrollBar {
+  XCTAssertEqual(scrollView.contentInsetAdjustmentBehavior,
+                 UIScrollViewContentInsetAdjustmentNever);
+  XCTAssertTrue(
+      UIEdgeInsetsEqualToEdgeInsets(scrollView.contentInset, UIEdgeInsetsMake(0, 0, 0, 0)));
```
```suggestion
      UIEdgeInsetsEqualToEdgeInsets(scrollView.contentInset, UIEdgeInsetsZero));
```

cbracken commented on shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:
```diff
@@ -154,6 +154,8 @@ - (instancetype)initWithBridge:(fml::WeakPtr<flutter::AccessibilityBridgeIos>)br
     _scrollView = [[FlutterSemanticsScrollView alloc] initWithSemanticsObject:self];
     [_scrollView setShowsHorizontalScrollIndicator:NO];
     [_scrollView setShowsVerticalScrollIndicator:NO];
+    [_scrollView setContentInset:UIEdgeInsetsMake(0, 0, 0, 0)];
```
```suggestion
    [_scrollView setContentInset:UIEdgeInsetsZero];
```

cbracken commented on shell/platform/darwin/ios/framework/Source/SemanticsObjectTest.mm:
```diff
@@ -1207,17 +1230,57 @@ - (void)testUIFocusItemContainerConformance {
+  // setContentOffset
+  CGPoint p = CGPointMake(123.0, 456.0);
+  [scrollable.scrollView scrollViewWillEndDragging:scrollable.scrollView
+                                      withVelocity:CGPointMake(0, 0)
```
```suggestion
                                      withVelocity:CGPointZero
```

cbracken commented on shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm:
```diff
@@ -87,16 +145,99 @@ - (CGRect)frame {
+///  `coordinateSpace` as the containing UIScrollView, or the root `FlutterView`, whichever is
+///  closer.
+///
+/// Seealso the `frame` method implementation.
```
One more!
```suggestion
/// See also the `frame` method implementation.
```

cbracken reviewed (approved):
Love it!

* LGTM + codeowners approval for embedder.h change.
* Some comments and nits on the iOS side. Shout when you want a second pass!

cbracken commented on shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm:
```diff
@@ -71,8 +81,56 @@ - (BOOL)canBecomeFocused {
+    if ([ancestor isKindOfClass:[FlutterScrollableSemanticsObject class]]) {
+      scrollView = ((FlutterScrollableSemanticsObject*)ancestor).scrollView;
+      break;
+    }
```
Rather than `if (...) { ...; break; } ...;` consider `if (...) { ... } else { ... }` here.

LongCatIsLooong replied:
Sorry I don't think I understand the suggested change, it's trying to find the closest scrollable ancestor so it needs to break out of the for loop?

cbracken replied:
Ugh yeah sorry. Too quick a look at this + lack of coffee = nonsense comment. This is fine :)

cbracken commented on shell/platform/darwin/ios/framework/Source/SemanticsObject+UIFocusSystem.mm:
```diff
@@ -87,16 +145,99 @@ - (CGRect)frame {
+      [reversedItems addObject:((FlutterScrollableSemanticsObject*)child).scrollView];
+    } else {
+      [reversedItems addObject:child];
+    }
```
Consider extracting something a static like this for readability:
```objc
static id<UIFocusItem>* GetFocusItemForNode(SemanticsObject* node) {
  if ([node isKindOfClass:[FlutterScrollableSemanticsObject class]]) {
    return ((FlutterScrollableSemanticsObject*)node).scrollView;
  }
  return node;
}
```

cbracken replied:
Even better!

LongCatIsLooong commented:
Alright the checks are all green (except the one that https://github.com/flutter/flutter/pull/159108 should fix) and I think I've addressed all comments. Let me know if I missed anything. Ready for re-review.

LongCatIsLooong commented:
Oops sorry didn't meant to request re-review from the Android codeowners since I didn't make any changes.

cbracken reviewed (approved):
![LGTM stamp from a Japanese personal seal](https://github.com/user-attachments/assets/c949029e-069a-4694-ba33-76f59283c8c7)

* lgtm/codeowners approval for embedder.h change
* lgtm for iOS changes

chunhtai reviewed (approved):
LGTM

LongCatIsLooong commented:
Reason for revert: https://github.com/flutter/flutter/issues/159456
